### PR TITLE
Add beatmap loading timeout to prevent runaway loading scenarios

### PR DESCRIPTION
**Pull request by smoogipoo**
I've set the timeout to a very lenient 10 seconds, just to cover for slower devices. Although this stuff should really complete within a second on a normal beatmap (e.g. centipede takes <120ms across all modes).

Can be tested with this beatmap: https://osu.ppy.sh/beatmapsets/367431#osu/805850

### CODE REVIEW

File: osu.Game/Beatmaps/IWorkingBeatmap.cs [hunk 1]
```diff
@@ -60,8 +61,9 @@ public interface IWorkingBeatmap
         /// </summary>
         /// <param name="ruleset">The <see cref="RulesetInfo"/> to create a playable <see cref="IBeatmap"/> for.</param>
         /// <param name="mods">The <see cref="Mod"/>s to apply to the <see cref="IBeatmap"/>.</param>
+        /// <param name="timeout">The loading timeout.</param>
```
Comment (peppy) [inline osu.Game/Beatmaps/IWorkingBeatmap.cs]:
```suggestion
        /// <param name="timeout">The maximum lenght in milliseconds to wait for load to complete. Defaults to 10,000ms.</param>
```

File: osu.Game/Beatmaps/WorkingBeatmap.cs [hunk 2]
```diff
@@ -83,55 +83,73 @@ protected virtual Track GetVirtualTrack(double emptyLength = 0)
...
+                processor?.PreProcess();
 
-            foreach (var mod in mods.OfType<IApplicableToHitObject>())
-            {
+                // Compute default values for hitobjects, including creating nested hitobjects in-case they're needed
                 foreach (var obj in converted.HitObjects)
-                    mod.ApplyToHitObject(obj);
-            }
+                {
+                    cancellationSource.Token.ThrowIfCancellationRequested();
```
Comment (peppy) [inline osu.Game/Beatmaps/WorkingBeatmap.cs]: sure we don't want to re-throw this as a `TimeoutException`? Seems a bit weird for callers to handle `OperationCancelledException` when they don't pass in a token.

**Review by peppy — CHANGES_REQUESTED** [review]

**Review by bdach — COMMENTED** [review]
Tested the cancellation path by dropping in a `Thread.Sleep` in one spot. Kind of eh that it shows two notifications (one from wedge, one from player):

image

Log excerpt:

```
2020-03-15 14:15:25 [error]: Could not load beatmap sucessfully!
2020-03-15 14:15:25 [error]: System.OperationCanceledException: The operation was canceled.
2020-03-15 14:15:25 [error]: at System.Threading.CancellationToken.ThrowOperationCanceledException()
2020-03-15 14:15:25 [error]: at System.Threading.CancellationToken.ThrowIfCancellationRequested()
2020-03-15 14:15:25 [error]: at osu.Game.Beatmaps.WorkingBeatmap.GetPlayableBeatmap(RulesetInfo ruleset, IReadOnlyList`1 mods, Nullable`1 timeout) in D:\Open Source\osu\osu.Game\Beatmaps\WorkingBeatmap.cs:line 132
2020-03-15 14:15:25 [error]: at osu.Game.Screens.Select.BeatmapInfoWedge.BufferedWedgeInfo.getInfoLabels() in D:\Open Source\osu\osu.Game\Screens\Select\BeatmapInfoWedge.cs:line 322
2020-03-15 14:15:25 [debug]: Focus changed from nothing to NotificationOverlay.
2020-03-15 14:15:27 [error]: Could not load beatmap sucessfully!
2020-03-15 14:15:27 [error]: System.OperationCanceledException: The operation was canceled.
2020-03-15 14:15:27 [error]: at System.Threading.CancellationToken.ThrowOperationCanceledException()
2020-03-15 14:15:27 [error]: at System.Threading.CancellationToken.ThrowIfCancellationRequested()
2020-03-15 14:15:27 [error]: at osu.Game.Beatmaps.WorkingBeatmap.GetPlayableBeatmap(RulesetInfo ruleset, IReadOnlyList`1 mods, Nullable`1 timeout) in D:\Open Source\osu\osu.Game\Beatmaps\WorkingBeatmap.cs:line 132
2020-03-15 14:15:27 [error]: at osu.Game.Screens.Play.Player.loadPlayableBeatmap() in D:\Open Source\osu\osu.Game\Screens\Play\Player.cs:line 318
```
Also besides @peppy's notes, a few comment / typo nits.

**Comment by smoogipoo** [comment]
@bdach going forward, we'll likely be getting the playable beatmap only once in a central location like OsuGameBase. The whole `GetPlayableBeatmap` process is too open. There's going to be other places that will require it (like recomputing scores on leaderboards), and having many playable beatmaps computed is wasteful.

**Review by bdach — APPROVED** [review]
looks good 👍  the double load definitely out of scope here

**Pull request merged**